From c6cd95a4d007437d9e5ae0afadca5e5e7ff49adb Mon Sep 17 00:00:00 2001
From: Manikanta Mylavarapu <mmanikan@qti.qualcomm.com>
Date: Sun, 14 Jun 2026 23:13:42 +0530
Subject: [PATCH] regulator: qcom: Add IPQ9650 open loop CPR driver config

Add Kconfig and Makefile entries for the Qualcomm IPQ9650 open loop
CPR regulator driver. The driver provides voltage regulation for the
APC by reading fuse values via NVMEM and calculating voltages for
different performance modes used during DVFS operations.

Change-Id: I6b088c337ffb3dc01c607772a82e6b143588baed
Signed-off-by: Manikanta Mylavarapu <mmanikan@qti.qualcomm.com>
---
 drivers/regulator/Kconfig  | 18 ++++++++++++++++++
 drivers/regulator/Makefile |  1 +
 2 files changed, 19 insertions(+)

diff --git a/drivers/regulator/Kconfig b/drivers/regulator/Kconfig
index d35ce57d2f6f..e7c40ac9815c 100644
--- a/drivers/regulator/Kconfig
+++ b/drivers/regulator/Kconfig
@@ -1090,6 +1090,24 @@ config REGULATOR_QCOM_USB_VBUS
 	  Say M here if you want to include support for enabling the VBUS output
 	  as a module. The module will be named "qcom_usb_vbus_regulator".
 
+config REGULATOR_QCOM_OPEN_LOOP_CPR
+	tristate "Qualcomm Open Loop CPR regulator support"
+	depends on REGULATOR
+	depends on (ARCH_QCOM || COMPILE_TEST)
+	depends on OF
+	depends on NVMEM
+	help
+	  This driver provides open loop Core Power Reduction (CPR) voltage
+	  regulation for Qualcomm IPQ9650 Application Processor Cluster (APC).
+	  It reads fuse correction values from eFuse registers via NVMEM and
+	  calculates the final operating voltages for SVS, NOM, and TUR modes.
+	  The calculated voltages are supplied through the regulator framework
+	  when invoked by the CPUFREQ framework during dynamic voltage and
+	  frequency scaling (DVFS) operations.
+
+	  Say M here to build this driver as a module. The module will be
+	  named "qcom-open-loop-cpr-regulator".
+
 config REGULATOR_RAA215300
 	tristate "Renesas RAA215300 driver"
 	select REGMAP_I2C
diff --git a/drivers/regulator/Makefile b/drivers/regulator/Makefile
index 5b8267c80108..4a04d1afe3ca 100644
--- a/drivers/regulator/Makefile
+++ b/drivers/regulator/Makefile
@@ -121,6 +121,7 @@ obj-$(CONFIG_REGULATOR_CPR3_NPU) += cpr3-npu-regulator.o
 obj-$(CONFIG_REGULATOR_CPR4_APSS) += cpr4-apss-regulator.o
 obj-$(CONFIG_REGULATOR_QCOM_USB_VBUS) += qcom_usb_vbus-regulator.o
 obj-$(CONFIG_REGULATOR_QTI_GPIO) += qcom-gpio-regulator.o
+obj-$(CONFIG_REGULATOR_QCOM_OPEN_LOOP_CPR) += qcom-open-loop-cpr-regulator.o
 obj-$(CONFIG_REGULATOR_PALMAS) += palmas-regulator.o
 obj-$(CONFIG_REGULATOR_PCA9450) += pca9450-regulator.o
 obj-$(CONFIG_REGULATOR_PF8X00) += pf8x00-regulator.o
-- 
2.34.1

